Unexport even more symbols

alaviss opened the pull request:
Hopefully I've tackled them all.

Symbols exported by an empty file created with `nim c --app:lib empty.nim`:
Before (50 exported symbols):
<details>

```
0000000000020800 B NTI__77mFvmsOLKik79ci2hXkHEg_
0000000000027ba0 B NTI__DsOOBcxCUeVlNUDRmn9afcA_
0000000000020740 B NTI__G9cUlLvU4AFC26wbFxLFkFA_
0000000000020700 B NTI__LbeSGvgPzGzXnW9caIkJqMA_
0000000000027b60 B NTI__Pjt0MQjoA6TAHAHOFNel9cg_
0000000000020d40 B NTI__S9agCYBinaYZnGWcjTdxclg_
0000000000020600 B NTI__ShBqCFAISBSH2YqBfe6zjg_
0000000000020780 B NTI__Ss6DFlX5iSZpHRZDmP74bg_
00000000000206c0 B NTI__Wyd9avMRCq0gsOu9adFoIjCA_
0000000000020860 B NTI__XEycrCsme5C8CVWAYEcdBQ_
0000000000020680 B NTI__XIT9aewsXycM2U5B437NUDA_
00000000000207c0 B NTI__hMQEc0FMry7Up7EoPki79aA_
0000000000020640 B NTI__oLyohQ7O2XOvGnflOss8EA_
0000000000020cc0 B NTI__rR5Bzr1D5krxoo1NcNyeMA_
0000000000027b20 B NTI__uB9b75OUPRENsBAu4AnoePA_
00000000000205c0 B NTI__vU9aO9cTqOMn6CBzhV8rX7Sw_
0000000000020d00 B NTI__ytyiCJqK439aF9cIibuRVpAg_
0000000000010080 T NimMain
00000000000100d0 T NimMainInit
0000000000010074 T NimMainInner
000000000001001f T PreMain
0000000000010018 T PreMainInner
0000000000011230 R TM__Q5wkpxktOdTGvlSRo9bzt9aw_11
0000000000011250 R TM__Q5wkpxktOdTGvlSRo9bzt9aw_12
0000000000010134 T _fini
0000000000002000 T _init
0000000000027bd8 B cmdCount
0000000000027be8 B cmdLine
0000000000020838 B currException__9bVPeDJlYTi9bQApZpfH8wjg
00000000000205a0 B errorMessageWriter__ZXnv0JyrWe3HTd07wpSr7A
0000000000020840 B excHandler__rqLlY5bs9atDw2OXYqJEn5g
0000000000020d78 B framePtr__HRfVMH3jYeBJz6Q6X9b6Ptw
0000000000011040 R fsLookupTable__Gn52IZvqY4slyBTOYwGNRQ
0000000000027be0 B gEnv
0000000000019838 B gHeapidGenerator__hd54mEUTGcVuZLChYgtR9bg
0000000000027b58 B gcFramePtr__ot48iojqko9aFxGhyjjjVaA
0000000000016460 B gch__IcYaEuuWivYAS86vFMTS3Q
0000000000020da0 B globalMarkers
0000000000020cf8 B globalMarkersLen
00000000000205f8 B globalRaiseHook__JbO1ti4ULxrw54m4zNPbpA
0000000000027b00 B localRaiseHook__EIvMhANBvB9cp2Ezvt29cADg
00000000000088e5 T markStackAndRegisters__U6T7JWtDLrWhtmhXSoy9a6g
0000000000027b08 B nim_program_result
0000000000020d80 B onUnhandledException__bFrawQlTKZhLweDD36j9b8g
0000000000020ca0 B outOfMemHook__kZNaA7u1MfSW5ZeoGvw8xg
0000000000016020 D strDesc__D0UzA4zsDu5tgpJQ9a9clXPg
00000000000208a0 B tempFrames__7nBYIr2UsDREpYylZK4fug
0000000000019840 B threadLocalMarkers
0000000000020d88 B threadLocalMarkersLen
0000000000019830 B unhandledExceptionHook__RJpSsXsRUH8ochbXTEhRIw
```
</details>

After (3 exported symbols):
<details>

```
000000000000ee4a T NimMain
000000000000eef4 T _fini
0000000000001000 T _init
```
</details>

alaviss committed:
system/gc: don't export markStackAndRegisters

alaviss committed:
compiler/cgen: unexport internal symbols

As these functions are Nim-specific walkaround against C's optimization
schemes, they don't serve any purpose being exported.

alaviss committed:
compiler/cgen: don't export global var unless marked

alaviss committed:
compiler/ccgthreadvars: don't export threadvar unless marked

alaviss committed:
tests/dll/visibility: also check for exports

This ensure that these changes don't break manual exports.

alaviss committed:
compiler/cgen: hide all variables created for constants

alaviss committed:
compiler/ccgtypes: don't export RTTI variables

alaviss committed:
compiler/ccgexprs: make all complex const static

juancarlospaco commented:
Needs a changelog entry.

I wonder if this breaks stuff like Nimpy and nBindGen. :thinking:

alaviss committed:
nimbase.h: fix export for windows

alaviss committed:
compiler/cgen, ccgthreadvars: export variables correctly

For C/C++ variables, `extern` means that the variable is defined in an
another unit. Added a new N_LIB_EXPORT_VAR to correctly export
variables.

alaviss commented:
> I wonder if this breaks stuff like Nimpy and nBindGen. :thinking:

It shouldn't. If something already works on Windows, it'll continue to work. All of this "hiding" only applies to *nix, whose default behavior is to just export everything unless marked (the opposite applies for Windows).

You can think of this PR as more or less enforcing the use of `{.dynlib.}` on *nix for symbols that you want to export. For Windows this PR fixes global variable export.
